immuni_analytics/helpers/request.py

```diff
@@ -0,0 +1,82 @@
+#   Copyright (C) 2020 Presidenza del Consiglio dei Ministri.
+#   Please refer to the AUTHORS file for more information.
+#   This program is free software: you can redistribute it and/or modify
+#   it under the terms of the GNU Affero General Public License as
+#   published by the Free Software Foundation, either version 3 of the
+#   License, or (at your option) any later version.
+#   This program is distributed in the hope that it will be useful,
+#   but WITHOUT ANY WARRANTY; without even the implied warranty of
+#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+#   GNU Affero General Public License for more details.
+#   You should have received a copy of the GNU Affero General Public License
+#   along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+import logging
+from typing import Any, Callable, Dict
+
+from aiohttp import ClientError, ClientSession, ClientTimeout
+from tenacity import (
+    RetryCallState,
+    retry,
+    retry_if_exception_type,
+    stop_after_attempt,
+    wait_exponential,
+)
+
+from immuni_common.core.exceptions import ImmuniException
+
+logger = logging.getLogger(__name__)
+
+
+class ServerUnavailableError(ImmuniException):
+    """Raised when the server returns a 5xx error code"""
+
+
+class BadFormatRequestError(ImmuniException):
+    """Raised when the server returns a 4xx error code"""
+
+
+def after_log() -> Callable[[RetryCallState], None]:
+    """After call strategy that logs the finished attempt"""
+
+    def log_it(retry_state: RetryCallState) -> None:
+        """Logs the error and the attempt"""
+        exc = retry_state.outcome.exception()
+        url = retry_state.kwargs.get("url")
+        logger.warning(
+            "Failed HTTP request",
+            url,
+            retry_state.attempt_number,
+            extra=dict(
+                exc=exc, request_args=retry_state, request_kwargs=retry_state.kwargs, url=url
+            ),
+        )
+
+    return log_it
+
+
+@retry(
+    retry=retry_if_exception_type((ClientError, TimeoutError, ServerUnavailableError)),
+    wait=wait_exponential(multiplier=1, min=2, max=10),
+    stop=stop_after_attempt(3),
+    after=after_log(),
```

Comment:
Why do you create a function that returns a function? Can't you just pass `log_it` (with a better name)?